migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2020 The Libphonenumber Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.phonenumbers.migrator;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.i18n.phonenumbers.metadata.RangeSpecification;
+import com.google.i18n.phonenumbers.metadata.RangeTree;
+import com.google.i18n.phonenumbers.metadata.i18n.PhoneRegion;
+import com.google.i18n.phonenumbers.metadata.table.CsvTable;
+import com.google.i18n.phonenumbers.metadata.table.RangeKey;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Scanner;
+
+/**
+ * Represents a migration operation for a given region where each {@link MigrationJob} contains
+ * a map of E.164 numbers to be migrated as well as the {@link CsvTable} which will
+ * hold the available recipes that can be performed on the range. The number range map is a key value
+ * pair of the E.164 {@link RangeSpecification} representation of a number along with the raw input
+ * String originally entered. Only recipes from the given two digit BCP-47 regionCode will be used.
+ */
+public final class MigrationJob {
+
+  private final CsvTable<RangeKey> recipesTable;
+  private final ImmutableMap<RangeSpecification, String> numberRangeMap;
+  private final PhoneRegion regionCode;
+
+  public MigrationJob(ImmutableMap<RangeSpecification,
+      String> numberRangeMap,
+      PhoneRegion regionCode,
+      CsvTable<RangeKey> recipesTable) {
+    this.numberRangeMap = numberRangeMap;
+    this.regionCode = regionCode;
+    this.recipesTable = recipesTable;
+  }
+
+  public CsvTable<RangeKey> getRecipesTable() {
+    return recipesTable;
+  }
+
+  public Map<RangeSpecification, String> getNumberRangeMap() {
```

Comment:
Best Practice: This is an interesting one.

I know that in Java/OO languages people say "return the generic type" (i.e. List<Foo> not ArrayList<Foo>) but in the case of immutable classes, it's almost always better to return the specific immutable type (i.e. ImmutableMap here).

This is because the "ImmutableFoo" classes actually limit functionality (e.g. the set/put methods don't work). Which is not clear to the caller if they get the generic type. Better than this, the Guava immutable collection classes use @deprecated annotation on mutable methods, so if you have an ImmutableMap and you call put(..) on it, the IDE will instantly cross out the method, making it clear there's an issue.

It's a hack, but a damn useful one.